### DISCUSSION TITLE
F-Secure USB Armory

### DISCUSSION THREAD

**Discussion by bernardoaraujor** [iotaledger/stronghold.rs#262] [Show and tell]
# Description

This issue tracks Stronghold integration with the F-Secure USB Armory.
Notes: 
- All the information contained on this page was collected from the official wiki and main project page.
- macOS commands might differ. Check official docs.

## Introduction

![image](https://inversepath.com/images/usbarmory_mark-two/usbarmory_coin_2.jpg)

> The USB armory board has been created to support the development of a variety of security applications.
> 
> The capability of emulating arbitrary USB devices in combination with the i.MX6UL SoC speed, the security features and the flexible and fully customizable operating environment, makes the USB armory the ideal platform for all kinds of personal security applications.
> 
> The transparency of the open and minimal design for the USB armory hardware facilitates auditability and greatly limits the potentiality and scope of supply chain attacks.
> 
> The secure boot feature allows users to fuse verification keys that ensure only trusted firmware can be ever executed on a specific USB armory board.
> 
> The support for ARM® TrustZone®, in contrast to conventional TPMs, allows developers to engineer custom trusted platform modules by enforcing domain separation, between the "secure" and "normal" worlds, that propagates throughout all SoC components, and therefore not only limited to the CPU core.
> 
> An excellent overview of the technology and its support for the i.MX6UL SoC can be found at the [Genode](https://genode.org/documentation/articles/trustzone) framework project. 

Any efforts in the direction of a Stronghold OS could laregely benefit from Genode’s information related to TEE.

## USB armory Mk II boot modes
The USB armory Mk II supports 3 boot mechanisms:

- internal 16GB eMMC
- external microSD

## USB Serial Download Protocol (SDP)

A slide switch allows selection of the primary boot mode. The switch positions match the position of the storage media components:
- left (towards eMMC): boot from internal eMMC
- right (towards microSD slot): boot from external microSD
![image](https://github.com/f-secure-foundry/usbarmory/wiki/images/armory-mark-two-boot-switch.png)

## Flashing bootable images on external/internal media

The following instructions provide instructions on flashing bootable images on either an external microSD card, connected to a host with a built-in or externally connected card reader, or the internal eMMC.

WARNING: the following operations will destroy any previous contents on the target eMMC/microSD.

## Available images
In order to boot the USB armory it is possible to either use a pre-compiled disk image or [manually prepare your own](https://github.com/f-secure-foundry/usbarmory/wiki/Preparing-a-bootable-image).

The following table summarizes the list of available pre-compiled Linux distributions

Device | OS | Author | Link
-- | -- | -- | --
Mk II | Debian 10 (Buster) | F-Secure Foundry | latest release
Mk II | Arch Linux | Arch Linux ARM | Arch Linux ARM Installation

Note: `*-usd-*` suffix on image filenames means `uSD`, or `microSD`, as opposed to `*-emmc-*` or `eMMC`.

However, **you should only trust what you can (cross)-compile**! Check this repo for instructions on how to cross-compile Debian: https://github.com/f-secure-foundry/usbarmory-debian-base_image

## Flashing raw disk images
Raw disk images are meant to fill the entire boot media, examples include full OS images such as the USB armory Debian base image.

Linux (verify target from terminal using dmesg, e.g. /dev/sdX):

```
$ sudo dd if=image.raw of=$TARGET_DEV bs=1M conv=fsync
```

## UART
The USB armory Mk II exposes a USB Type-C receptacle, which allows through the so called USB-C 'debug accessory mode' to route analog/debug signals over its connector, the USB armory Mk II design leverages on this to break out UART, SPI, I²C, CAN (pre-transceiver), GPIOs connections to/from its application processor.

![image](https://github.com/f-secure-foundry/usbarmory/wiki/images/armory-mark-two-debug-accessory.png)

The debug accessory board allows to access UART and GPIO signals through USB, without requiring probes, through an FTDI FT4232H. This allows, for example, access to the USB armory Mk II serial console without wires or probes, natively using only USB cables.

When running the USB armory Debian base image the serial console can be accessed, with the debug accessory jumper set to UART position, as follows (example on Linux host with picocom):

```
$ sudo picocom -b 115200 -eb /dev/ttyUSB2 --imap lfcrlf
``` 

When prompted for username and password, use `usbarmory` for both.

**WARNING**: make sure the Jumper is setting the adapter to UART mode.

# Ethernet over USB (SSH)
After being booted, the image uses Ethernet over USB emulation (CDC Ethernet) to communicate with the host, with assigned IP address `10.0.0.1` (using `10.0.0.2` as gateway). Connection can be accomplished via **SSH** to `10.0.0.1`, with default user `usbarmory` and password `usbarmory`. NOTE: There is a DHCP server running by default. Alternatively the host interface IP address can be statically set to `10.0.0.2/24`.

```
$ ssh usbarmory@10.0.0.1
```

**Comment by elenaf9** [comment]
# Host communication
Setup & Connection Sharing  Linux:
(please replace usb0 and wlan0 with the correct identifiers from interface configuration)
```
# bring the USB virtual Ethernet interface up
/sbin/ip link set usb0 up

# set the host IP address
/sbin/ip addr add 10.0.0.2/24 dev usb0

# enable masquerading for outgoing connections towards wireless interface
/sbin/iptables -t nat -A POSTROUTING -s 10.0.0.1/32 -o wlan0 -j MASQUERADE

# enable IP forwarding
echo 1 > /proc/sys/net/ipv4/ip_forward

```
Additionally, on Ubuntu the firewall has to be configured to allow the traffic:
```
iptables -A FORWARD -s 10.0.0.1 -j ACCEPT
```
and in some cases it is necessary to also restart the network-manager.

# Compiling Rust for the USB Armory Mk II

The USB Armory hosts per default the `stable-arm-unknown-linux-gnueabihf ` toolchain, thus a rust project has to be compiled with `--target=arm-unknown-linux-gnueabihf`. The linker is configured in the`.cargo/config` to use `arm-linux-gnueabihf-gcc` (#90), which has to be installed beforehand.
Additionally, the crate `ring v0.16.19` requires the environment variable `TARGET_CC=arm-linux-gnueabihf-gcc`.
```
TARGET_CC=arm-linux-gnueabihf-gcc cargo build --target=arm-unknown-linux-gnueabihf
```
I compiled the binary on my host Ubuntu laptop and copied it to the usb-armory. Compiling on the usb armory itself takes a few hours and in my case the linker is always killed, seems like it costs too much ram.
These steps worked for the stronghold-communication package so far and enabled the communication between the host computer and the usb-armory. 
However there are still issues when compiling the whole stronghold workspace:
```
 Compiling snapshot v0.1.0 (/stronghold.rs/engine/snapshot)
   Compiling stronghold-runtime v0.1.0 (/stronghold.rs/runtime)
   Compiling stronghold-engine v0.1.0 (/stronghold.rs/engine)
error: failed to run custom build command for `stronghold-runtime v0.1.0 (/stronghold.rs/runtime)`

Caused by:
  process didn't exit successfully: `/stronghold.rs/target/debug/build/stronghold-runtime-51f071db0e0cf9c2/build-script-build` (exit code: 101)
  --- stdout
  cargo:rerun-if-changed=src/seccomp.h

  --- stderr
  /usr/include/linux/types.h:5:10: fatal error: 'asm/types.h' file not found
  /usr/include/linux/types.h:5:10: fatal error: 'asm/types.h' file not found, err: true
  thread 'main' panicked at 'Unable to generate bindings: ()', runtime/build.rs:17:10
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
error: build failed
```

**Comment by elenaf9** [comment]
The previous problem with compiling the stronghold-runtime for the `target=arm-unknown-linux-gnueabihf` can be solved by setting the environment variable `C_INCLUDE_PATH` to the path of `arm-linux-gnueabihf/include`. 
On Ubuntu this would be `C_INCLUDE_PATH=/usr/arm-linux-gnueabihf/include`, this directory should exist after the previous installation of `arm-linux-gnueabihf-gcc`.
